// @mui/x-data-grid-pro/node/components/GridScrollArea.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridScrollArea = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _utils = require("@mui/utils");
var _system = require("@mui/system");
var _internals = require("@mui/x-data-grid/internals");
var _xDataGrid = require("@mui/x-data-grid");
var _useGridRootProps = require("../hooks/utils/useGridRootProps");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const CLIFF = 1;
const SLOP = 1.5;
const useUtilityClasses = ownerState => {
  const {
    scrollDirection,
    classes
  } = ownerState;
  const slots = {
    root: ['scrollArea', `scrollArea--${scrollDirection}`]
  };
  return (0, _utils.unstable_composeClasses)(slots, _xDataGrid.getDataGridUtilityClass, classes);
};
const GridScrollAreaRawRoot = (0, _system.styled)('div', {
  name: 'MuiDataGrid',
  slot: 'ScrollArea',
  overridesResolver: (props, styles) => [{
    [`&.${_xDataGrid.gridClasses['scrollArea--left']}`]: styles['scrollArea--left']
  }, {
    [`&.${_xDataGrid.gridClasses['scrollArea--right']}`]: styles['scrollArea--right']
  }, styles.scrollArea]
})(() => ({
  position: 'absolute',
  top: 0,
  zIndex: 101,
  width: 20,
  bottom: 0,
  [`&.${_xDataGrid.gridClasses['scrollArea--left']}`]: {
    left: 0
  },
  [`&.${_xDataGrid.gridClasses['scrollArea--right']}`]: {
    right: 0
  }
}));
function GridScrollAreaRaw(props) {
  const {
    scrollDirection
  } = props;
  const rootRef = React.useRef(null);
  const apiRef = (0, _xDataGrid.useGridApiContext)();
  const timeout = (0, _internals.useTimeout)();
  const [dragging, setDragging] = React.useState(false);
  const [canScrollMore, setCanScrollMore] = React.useState(true);
  const densityFactor = (0, _xDataGrid.useGridSelector)(apiRef, _xDataGrid.gridDensityFactorSelector);
  const columnsTotalWidth = (0, _xDataGrid.useGridSelector)(apiRef, _xDataGrid.gridColumnsTotalWidthSelector);
  const scrollPosition = React.useRef({
    left: 0,
    top: 0
  });
  const rootProps = (0, _useGridRootProps.useGridRootProps)();
  const ownerState = (0, _extends2.default)({}, rootProps, {
    scrollDirection
  });
  const classes = useUtilityClasses(ownerState);
  const totalHeaderHeight = (0, _internals.getTotalHeaderHeight)(apiRef, rootProps.columnHeaderHeight);
  const headerHeight = Math.floor(rootProps.columnHeaderHeight * densityFactor);
  const handleScrolling = React.useCallback(newScrollPosition => {
    scrollPosition.current = newScrollPosition;
    const dimensions = apiRef.current.getRootDimensions();
    setCanScrollMore(() => {
      if (scrollDirection === 'left') {
        // Only render if the user has not reached yet the start of the list
        return scrollPosition.current.left > 0;
      }
      if (scrollDirection === 'right') {
        // Only render if the user has not reached yet the end of the list
        const maxScrollLeft = columnsTotalWidth - dimensions.viewportInnerSize.width;
        return scrollPosition.current.left < maxScrollLeft;
      }
      return false;
    });
  }, [apiRef, columnsTotalWidth, scrollDirection]);
  const handleDragOver = React.useCallback(event => {
    let offset;

    // Prevents showing the forbidden cursor
    event.preventDefault();
    if (scrollDirection === 'left') {
      offset = event.clientX - rootRef.current.getBoundingClientRect().right;
    } else if (scrollDirection === 'right') {
      offset = Math.max(1, event.clientX - rootRef.current.getBoundingClientRect().left);
    } else {
      throw new Error('MUI: Wrong drag direction');
    }
    offset = (offset - CLIFF) * SLOP + CLIFF;

    // Avoid freeze and inertia.
    timeout.start(0, () => {
      apiRef.current.scroll({
        left: scrollPosition.current.left + offset,
        top: scrollPosition.current.top
      });
    });
  }, [scrollDirection, apiRef, timeout]);
  const handleColumnHeaderDragStart = (0, _utils.unstable_useEventCallback)(() => {
    setDragging(true);
  });
  const handleColumnHeaderDragEnd = (0, _utils.unstable_useEventCallback)(() => {
    setDragging(false);
  });
  (0, _xDataGrid.useGridApiEventHandler)(apiRef, 'scrollPositionChange', handleScrolling);
  (0, _xDataGrid.useGridApiEventHandler)(apiRef, 'columnHeaderDragStart', handleColumnHeaderDragStart);
  (0, _xDataGrid.useGridApiEventHandler)(apiRef, 'columnHeaderDragEnd', handleColumnHeaderDragEnd);
  if (!dragging || !canScrollMore) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(GridScrollAreaRawRoot, {
    ref: rootRef,
    className: (0, _clsx.default)(classes.root),
    ownerState: ownerState,
    onDragOver: handleDragOver,
    style: {
      height: headerHeight,
      top: totalHeaderHeight - headerHeight
    }
  });
}
process.env.NODE_ENV !== "production" ? GridScrollAreaRaw.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------
  scrollDirection: _propTypes.default.oneOf(['left', 'right']).isRequired
} : void 0;
const GridScrollArea = exports.GridScrollArea = /*#__PURE__*/React.memo(GridScrollAreaRaw);
