// @mui/x-data-grid-pro/legacy/components/GridScrollArea.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { unstable_composeClasses as composeClasses, unstable_useEventCallback as useEventCallback } from '@mui/utils';
import { styled } from '@mui/system';
import { getTotalHeaderHeight, useTimeout } from '@mui/x-data-grid/internals';
import { getDataGridUtilityClass, gridClasses, gridDensityFactorSelector, useGridApiContext, useGridApiEventHandler, useGridSelector, gridColumnsTotalWidthSelector } from '@mui/x-data-grid';
import { useGridRootProps } from '../hooks/utils/useGridRootProps';
import { jsx as _jsx } from "react/jsx-runtime";
var CLIFF = 1;
var SLOP = 1.5;
var useUtilityClasses = function useUtilityClasses(ownerState) {
  var scrollDirection = ownerState.scrollDirection,
    classes = ownerState.classes;
  var slots = {
    root: ['scrollArea', "scrollArea--".concat(scrollDirection)]
  };
  return composeClasses(slots, getDataGridUtilityClass, classes);
};
var GridScrollAreaRawRoot = styled('div', {
  name: 'MuiDataGrid',
  slot: 'ScrollArea',
  overridesResolver: function overridesResolver(props, styles) {
    return [_defineProperty({}, "&.".concat(gridClasses['scrollArea--left']), styles['scrollArea--left']), _defineProperty({}, "&.".concat(gridClasses['scrollArea--right']), styles['scrollArea--right']), styles.scrollArea];
  }
})(function () {
  return _defineProperty(_defineProperty({
    position: 'absolute',
    top: 0,
    zIndex: 101,
    width: 20,
    bottom: 0
  }, "&.".concat(gridClasses['scrollArea--left']), {
    left: 0
  }), "&.".concat(gridClasses['scrollArea--right']), {
    right: 0
  });
});
function GridScrollAreaRaw(props) {
  var scrollDirection = props.scrollDirection;
  var rootRef = React.useRef(null);
  var apiRef = useGridApiContext();
  var timeout = useTimeout();
  var _React$useState = React.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    dragging = _React$useState2[0],
    setDragging = _React$useState2[1];
  var _React$useState3 = React.useState(true),
    _React$useState4 = _slicedToArray(_React$useState3, 2),
    canScrollMore = _React$useState4[0],
    setCanScrollMore = _React$useState4[1];
  var densityFactor = useGridSelector(apiRef, gridDensityFactorSelector);
  var columnsTotalWidth = useGridSelector(apiRef, gridColumnsTotalWidthSelector);
  var scrollPosition = React.useRef({
    left: 0,
    top: 0
  });
  var rootProps = useGridRootProps();
  var ownerState = _extends({}, rootProps, {
    scrollDirection: scrollDirection
  });
  var classes = useUtilityClasses(ownerState);
  var totalHeaderHeight = getTotalHeaderHeight(apiRef, rootProps.columnHeaderHeight);
  var headerHeight = Math.floor(rootProps.columnHeaderHeight * densityFactor);
  var handleScrolling = React.useCallback(function (newScrollPosition) {
    scrollPosition.current = newScrollPosition;
    var dimensions = apiRef.current.getRootDimensions();
    setCanScrollMore(function () {
      if (scrollDirection === 'left') {
        // Only render if the user has not reached yet the start of the list
        return scrollPosition.current.left > 0;
      }
      if (scrollDirection === 'right') {
        // Only render if the user has not reached yet the end of the list
        var maxScrollLeft = columnsTotalWidth - dimensions.viewportInnerSize.width;
        return scrollPosition.current.left < maxScrollLeft;
      }
      return false;
    });
  }, [apiRef, columnsTotalWidth, scrollDirection]);
  var handleDragOver = React.useCallback(function (event) {
    var offset;

    // Prevents showing the forbidden cursor
    event.preventDefault();
    if (scrollDirection === 'left') {
      offset = event.clientX - rootRef.current.getBoundingClientRect().right;
    } else if (scrollDirection === 'right') {
      offset = Math.max(1, event.clientX - rootRef.current.getBoundingClientRect().left);
    } else {
      throw new Error('MUI: Wrong drag direction');
    }
    offset = (offset - CLIFF) * SLOP + CLIFF;

    // Avoid freeze and inertia.
    timeout.start(0, function () {
      apiRef.current.scroll({
        left: scrollPosition.current.left + offset,
        top: scrollPosition.current.top
      });
    });
  }, [scrollDirection, apiRef, timeout]);
  var handleColumnHeaderDragStart = useEventCallback(function () {
    setDragging(true);
  });
  var handleColumnHeaderDragEnd = useEventCallback(function () {
    setDragging(false);
  });
  useGridApiEventHandler(apiRef, 'scrollPositionChange', handleScrolling);
  useGridApiEventHandler(apiRef, 'columnHeaderDragStart', handleColumnHeaderDragStart);
  useGridApiEventHandler(apiRef, 'columnHeaderDragEnd', handleColumnHeaderDragEnd);
  if (!dragging || !canScrollMore) {
    return null;
  }
  return /*#__PURE__*/_jsx(GridScrollAreaRawRoot, {
    ref: rootRef,
    className: clsx(classes.root),
    ownerState: ownerState,
    onDragOver: handleDragOver,
    style: {
      height: headerHeight,
      top: totalHeaderHeight - headerHeight
    }
  });
}
process.env.NODE_ENV !== "production" ? GridScrollAreaRaw.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------
  scrollDirection: PropTypes.oneOf(['left', 'right']).isRequired
} : void 0;
var GridScrollArea = /*#__PURE__*/React.memo(GridScrollAreaRaw);
export { GridScrollArea };
