// @mui/x-data-grid-pro/legacy/hooks/features/treeData/gridTreeDataUtils.js
import { passFilterLogic } from '@mui/x-data-grid/internals';
export var TREE_DATA_STRATEGY = 'tree-data';

/**
 * A node is visible if one of the following criteria is met:
 * - One of its children is passing the filter
 * - It is passing the filter
 */
export var filterRowTreeFromTreeData = function filterRowTreeFromTreeData(params) {
  var apiRef = params.apiRef,
    rowTree = params.rowTree,
    disableChildrenFiltering = params.disableChildrenFiltering,
    isRowMatchingFilters = params.isRowMatchingFilters;
  var filteredRowsLookup = {};
  var filteredDescendantCountLookup = {};
  var filterCache = {};
  var filterResults = {
    passingFilterItems: null,
    passingQuickFilterValues: null
  };
  var filterTreeNode = function filterTreeNode(node, isParentMatchingFilters, areAncestorsExpanded) {
    var shouldSkipFilters = disableChildrenFiltering && node.depth > 0;
    var isMatchingFilters;
    if (shouldSkipFilters) {
      isMatchingFilters = null;
    } else if (!isRowMatchingFilters || node.type === 'footer') {
      isMatchingFilters = true;
    } else {
      var row = apiRef.current.getRow(node.id);
      isRowMatchingFilters(row, undefined, filterResults);
      isMatchingFilters = passFilterLogic([filterResults.passingFilterItems], [filterResults.passingQuickFilterValues], params.filterModel, params.apiRef, filterCache);
    }
    var filteredDescendantCount = 0;
    if (node.type === 'group') {
      node.children.forEach(function (childId) {
        var _isMatchingFilters;
        var childNode = rowTree[childId];
        var childSubTreeSize = filterTreeNode(childNode, (_isMatchingFilters = isMatchingFilters) != null ? _isMatchingFilters : isParentMatchingFilters, areAncestorsExpanded && !!node.childrenExpanded);
        filteredDescendantCount += childSubTreeSize;
      });
    }
    var shouldPassFilters;
    switch (isMatchingFilters) {
      case true:
        {
          shouldPassFilters = true;
          break;
        }
      case false:
        {
          shouldPassFilters = filteredDescendantCount > 0;
          break;
        }
      default:
        {
          shouldPassFilters = isParentMatchingFilters;
          break;
        }
    }
    filteredRowsLookup[node.id] = shouldPassFilters;
    if (!shouldPassFilters) {
      return 0;
    }
    filteredDescendantCountLookup[node.id] = filteredDescendantCount;
    if (node.type === 'footer') {
      return filteredDescendantCount;
    }
    return filteredDescendantCount + 1;
  };
  var nodes = Object.values(rowTree);
  for (var i = 0; i < nodes.length; i += 1) {
    var node = nodes[i];
    if (node.depth === 0) {
      filterTreeNode(node, true, true);
    }
  }
  return {
    filteredRowsLookup: filteredRowsLookup,
    filteredDescendantCountLookup: filteredDescendantCountLookup
  };
};
